numba/core/typing/new_builtins.py

```diff
@@ -0,0 +1,1178 @@
+@infer_global(prange, typing_key=prange)
+@infer_global(internal_prange, typing_key=internal_prange)
+class Range(ConcreteTemplate):
+    cases = [
+        signature(types.range_state64_type, types.py_int64),
+        signature(types.range_state64_type, types.py_int64, types.py_int64),
+        signature(types.range_state64_type, types.py_int64, types.py_int64,
+                types.py_int64),
+    ]
+
+
+@infer
+class GetIter(AbstractTemplate):
+    key = "getiter"
+
+    def generic(self, args, kws):
+        assert not kws
+        [obj] = args
+        if isinstance(obj, types.IterableType):
+            return signature(obj.iterator_type, obj)
+
+
+@infer
+class IterNext(AbstractTemplate):
+    key = "iternext"
+
+    def generic(self, args, kws):
+        assert not kws
+        [it] = args
+        if isinstance(it, types.IteratorType):
+            return signature(types.Pair(it.yield_type, types.np_bool_), it)
+
+
+@infer
+class PairFirst(AbstractTemplate):
+    """
+    Given a heterogeneous pair, return the first element.
+    """
+    key = "pair_first"
+
+    def generic(self, args, kws):
+        assert not kws
+        [pair] = args
+        if isinstance(pair, types.Pair):
+            return signature(pair.first_type, pair)
+
+
+@infer
+class PairSecond(AbstractTemplate):
+    """
+    Given a heterogeneous pair, return the second element.
+    """
+    key = "pair_second"
+
+    def generic(self, args, kws):
+        assert not kws
+        [pair] = args
+        if isinstance(pair, types.Pair):
+            return signature(pair.second_type, pair)
+
+
+def choose_result_bitwidth(*inputs):
+    return max(tp.bitwidth for tp in inputs)
+
+def choose_result_int(*inputs):
+    """
+    Choose the integer result type for an operation on integer inputs,
+    according to the integer typing NBEP. In accordance with the new
+    type system.
+    """
+    bitwidth = choose_result_bitwidth(*inputs)
+    signed = any(tp.signed for tp in inputs)
+
+    # If any integer is a NumPy integer, promotion should be to the
+    # respective NumPy type.
+    if any('np' in tp.name for tp in inputs):
+        return types.NumPyInteger.from_bitwidth(bitwidth, signed)
+
+    return types.PythonInteger.from_bitwidth(bitwidth, signed)
+
+all_ints = (
+    sorted(set((types.py_intp, types.py_int64))) +
+    sorted(set((types.np_int32, types.np_int64))) +
+    sorted(set((types.np_uint32, types.np_uint64)))
+    )
+integer_binop_cases = tuple(
+    signature(choose_result_int(op1, op2), op1, op2)
+    for op1, op2 in itertools.product(all_ints, all_ints)
+    )
+
+class BinOp(ConcreteTemplate):
+    cases = list(integer_binop_cases)
+
+    cases += [signature(op, op, op) for op in sorted(types.py_real_domain)]
+    cases += [signature(op, op, op) for op in sorted(types.py_complex_domain)]
+    cases += [signature(op, op, op) for op in sorted(types.np_real_domain)]
+    cases += [signature(op, op, op) for op in sorted(types.np_complex_domain)]
```

Comment:
With the addition of protocol based `operator` implementations, as in #9669, the correctness of this (and all following templates) is not of concern as it will be deleted.